Indexed Entity X-Rays

markbastian opened the pull request:
Adds the ability to x-ray indexed entities. Primary logic is in `metabase.api.automagic-dashboards` and does the following:

- Adds an endpoint `GET "/model_index/:model-index-id/primary_key/:pk-id"` that takes the index and primary key of an indexed entity (i.e. a row in a table or model that has been indexed)
- Looks up the corresponding `ModelIndex`,`Card` (model), and `ModelIndexValue`
- Computes entities linked to this row via FK
- Generates a dashboard for each of these entities
- Merges the above dashboards into a single dashboard with each individual dashboard grouped onto tabs

markbastian committed:
metabase.automagic-dashboards.core - Docs, cleanup, and destructuring for clarity
metabase.api.search - Adding fix to realize search results
metabase.automagic-dashboards.filters - Fix to build-fk-map so that parameters show up on model x-rays
metabase.automagic-dashboards.populate - Fix typo
metabase.api.search-test - Unit tests for search fix

markbastian committed:
Brought over tab saving of transient x-ray dashboard code to save-transient-dashboard!

Added initial cut of tabbed x-ray generation from mapping over linked entities to the automagic-dashboards api

dpsutton committed:
update signature

we'll need some filtering criteria for the tabs. we don't just want an
x-ray of orders, we want an x-ray of _Customer's orders_

markbastian committed:
Added query-filter to /model_index/:model-index-id/primary_key/:pk-id API

markbastian committed:
Fixed id propagation

dpsutton committed:
clean up the logic a little bit

markbastian committed:
Intermediate step to trap bad generated queries

markbastian committed:
Merge remote-tracking branch 'origin/tabbed-x-rays' into tabbed-x-rays

dpsutton committed:
First stab at queries with joins for x-rays

take a model, find joinable tables and create joins back to the model's
underlying table so we can filter where pk=selected-id.

```clojure
{:database_id 1,
 :database-id 1,
 :table_id 2,
 :dataset_query {:database 1,
                 :type :query,
                 :query {:source-table 2,
                         :where [:=
                                 [:field
                                  62
                                  {:join-alias "ORDERS__via__PRODUCT_ID"}]
                                 1],
                         :joins [{:alias "ORDERS__via__PRODUCT_ID",
                                  :fields :none,
                                  :strategy :left-join,
                                  :condition [:=
                                              [:field
                                               62
                                               {:join-alias "ORDERS__via__PRODUCT_ID"}]
                                              [:field 40 nil]],
                                  :source-table 1}]}}}
```

Note the where clause and the join, but selecting `:fields :none` from
the join. so it's

```sql
select j.*
from joinable j
left join model m
where m.id = <user-selection>
```

But x-rays of these queries strip out our joins :(

Here's what comes back in the automagic dashboard

```clojure
{:type :query,
 :database 1,
 :query {:source-table 2,
         :filter [:and
                  [:=
                   [:field
                    62
                    nil]
                   1]],
         :breakout ([:field
                     41
                     {:temporal-unit :month}]),
         :aggregation (["avg"
                        [:field
                         39
                         nil]])}}
```

It's a breakout by month of orders.created_at (field 41) aggregated by
orders.quantity (field 39) where products.id (field 62 of table) =
1. But there's no join to products any longer :(

And there's this nice little comment

```clojure
(defn- maybe-enrich-joins "Hack to shove back in joins when they get
automagically stripped out by the question decomposition into metrics"
```

So that's no great. More investigating to do

markbastian committed:
Added in `:source-field` for our x-ray tabs so that filters would succeed.

markbastian committed:
Merge remote-tracking branch 'origin/tabbed-x-rays' into tabbed-x-rays

# Conflicts:
#	src/metabase/api/automagic_dashboards.clj

markbastian committed:
Merged fixes to filters.

markbastian committed:
Adding in entity x-ray dashboard title

markbastian committed:
Fixed entity x-ray description.

markbastian committed:
Added model-link card to each tab in the dynamically generated entity x-rays.

markbastian committed:
Adding min-width calculation for source model link card.

markbastian committed:
Adding tests for min-width calculation for source model link card.

dpsutton committed:
tests for x-rays on indexed models

markbastian committed:
Fixing import indentation

markbastian committed:
Adding tests to metabase.api.automagic-dashboards-test to assert the existence of model link cards and meaningful dashboard names and descriptions.

dpsutton committed:
include model_index_id in search results for indexed entity

had the rest of the changes, just not the id of the model_index

dpsutton committed:
use negative integers, not gensyms for tab-ids

FE wants negative integers for ids that do not exist in the db yet. some
of the cards and such still use gensym and that seems to be ok for now.

EmmadUsmani committed:
fix tabs on x-ray dashboard in FE

EmmadUsmani committed:
add x-ray button to `SearchResult` component

EmmadUsmani committed:
clean up styles for tabs in `AutomaticDashboardAppInner`

markbastian committed:
Added tests to understand some of the logic in `field-candidates.` The selection of a candidate is variable depending on the specification of the dimension, the field type, the table type, how the context is populated, etc. These tests attempt to illustrate some of the logic.

markbastian committed:
Adding tests in which multiple fields match to multiple dimensions.

markbastian committed:
Added tests for bind-dimensions and much of the logic it calls.

markbastian committed:
Removing `:transient_name` and `:transient_filters` from generated entity x-ray to ensure the right information is displayed on the generated dashboard.

markbastian committed:
Merge branch 'master' into tabbed-x-rays

markbastian commented on src/metabase/automagic_dashboards/filters.clj:
```diff
@@ -78,7 +78,7 @@
 
 (defn- build-fk-map
   [fks field]
```
This is a bugfix that ensures that filters show up on entities with no fks.

markbastian committed:
Formatting

markbastian committed:
Formatting

markbastian committed:
Formatting

markbastian committed:
Formatting

markbastian committed:
Merge branch 'master' into tabbed-x-rays

markbastian committed:
Fixing quoted instance of Card

markbastian committed:
Merge remote-tracking branch 'origin/tabbed-x-rays' into tabbed-x-rays

markbastian committed:
wslint

dpsutton committed:
Hook for `metabase.api.automagic-dashboards-test/with-indexed-model`

markbastian committed:
wslint

markbastian committed:
Merge remote-tracking branch 'origin/tabbed-x-rays' into tabbed-x-rays

markbastian committed:
Merge branch 'master' into tabbed-x-rays

dpsutton commented:
I think there's a rough edge that we need to smooth out. For mbql queries, we will often find linked tables to x-ray, but not always. And for native queries, we will never find linked tables to x-ray unless they have identified the pk column with a column in the database.

<img width="814" alt="image" src="https://github.com/metabase/metabase/assets/6377293/6b07048e-f383-4250-ae3b-53ef8ab6e766">

When this scenario happens, we don't have anything to x-ray and we just return an error. We need to make sure this error is a polished state in the flow, and it includes enough information to understand why it happened, and what they can do to correctly link up their model.

dpsutton commented:
For instance, I've created a native model on products (`select * from products`) named "native model on products without column info". Add an index marking the title as the label and id as the primary key.

<img width="1058" alt="image" src="https://github.com/metabase/metabase/assets/6377293/53800eba-3840-4574-9510-b31242a1cb6c">

When I go to the indexed x-ray I see the following:
<img width="1056" alt="image" src="https://github.com/metabase/metabase/assets/6377293/353385a8-12c2-4eb5-b5db-23f4e2c7e19f">

Not ideal.

dpsutton committed:
lint warnings on ns requires

unused ns qp
using a deprecated namespace so give an override

dpsutton committed:
better error message for no linked entities

they used to be in the let-404 so you would get a quite confusing error
message. Now it's a 400 and will show "No linked entities" if you expand
the error message. We still need to make this better. But this is an
obvious improvement on a 404.

markbastian committed:
Adding default entity dashboard for when no linked entities are found.

markbastian committed:
Merge remote-tracking branch 'origin/tabbed-x-rays' into tabbed-x-rays

# Conflicts:
#	src/metabase/api/automagic_dashboards.clj

markbastian commented:
I've added the following to address the no linked entities edge case.

![boring](https://github.com/metabase/metabase/assets/8381441/a2068567-bb74-4127-a22a-2037927e5eb5)

markbastian commented:
Here's another weird case in which an index is added to native model (`SELECT * FROM PEOPLE`) and an index is added on address.

![whatsup](https://github.com/metabase/metabase/assets/8381441/d374bd1d-1918-4b29-a5dc-91e6d7ca7da6)

dpsutton committed:
fix search tests

dpsutton committed:
call count is 16 now

markbastian committed:
Fixing unit test

markbastian committed:
Merge remote-tracking branch 'origin/tabbed-x-rays' into tabbed-x-rays

markbastian committed:
Merge branch 'master' into tabbed-x-rays

markbastian committed:
Fixed the query filter in `metabase.api.automagic-dashboards` so that `create-linked-dashboard` doe not produce bad queries.

dpsutton committed:
remove model index destructure

markbastian committed:
Merge branch 'master' into tabbed-x-rays

dpsutton committed:
Assert that all query dashcards have the correct filter.

We're no longer making joins back to the model's underlying
table. Recognizing that we were joining on linked table's fk to the
model's underlying table's pk and then searching where pk = <pk-value>,
we can just filter where fk = <pk-value>, omit the join.

So these tests just grab all of the linked tables and assert that one of
those filters is found.

eg, suppose a model based on products. the linked tables are orders and
reviews. And rather than the queries:

```sql
select o.*
from orders o
left join products p
on p.id = o.product_id
where p.id = <pk-value>
```

we can just do

```sql
select *
from orders
where product_id = <pk-value>
```

And similar for reviews. And for each query in the dashboard we're
looking for one of the following:

```clojure
, #{[:= $orders.product_id 1] [:= $reviews.product_id 1]}
```

markbastian committed:
Intermediate stage of fixing our tests

markbastian committed:
Added missing `is` clauses to a few tests. This surfaced some minor formatting changes that would have broken the tests had they been working. Updated the tests to correspond to the changes (just has to do with pluralization).

markbastian committed:
Fixed untyped event and added mock search data.

markbastian committed:
Fixing entity x-rays for a single linked entity

markbastian committed:
Adding logic to provide only a single linked entity view for entity x-rays if only one is generated.

markbastian committed:
lint fix

dpsutton committed:
add tests for 1 linked-table = 0 tabs case

markbastian commented:
There were a couple issues causing the spinner issues.

- Complex 2-step joins that can actually be accomplished in a single filter
- Using tab logic when there is only one tab, which causes FE rendering problems

Both of these have been fixed. Here's an example of a chart that was having issues 
with both cases and is now fixed.

![fixed](https://github.com/metabase/metabase/assets/8381441/e53122a1-ce41-4070-88fb-28feb1d18f91)

markbastian committed:
Updated `create-linked-dashboard` with a threading macro for clarity and to be more clojuric.

markbastian committed:
Merge branch 'master' into tabbed-x-rays

dpsutton committed:
split single test into three

cam made a new linter to look for test forms over a certain
size. There's no great way to run only a single testing form so this
helps make things more atomic

markbastian committed:
Merge branch 'master' into tabbed-x-rays

dpsutton committed:
Handle expression refs in indexed-models

Custom columns look like the following:

```clojure
{:expressions {"full-name" [:concat <first-name> <last-name>]}
```

To index values, we need a sequence of primary key and the associated
text value. So we select them from a nested query on the model. But the
model's way to refer to an expression and queries _on_ the model are
different. To the model, it's an expression. To queries based on the
model, it's just another field.

And have field refs in the result_metadata `[:expression
"full-name"]`. But when selecting from a nested query, they are just
string based fields: `[:field "full-name" {:base-type :type/Text}]`.

old style query we issued when fetching values:

```clojure
{:database 122
 :type :query
 :query {:source-table "card__1715"
         :breakout [[:field 488 {:base-type :type/Integer}]
                    [:expression "full name"]] ;; <-- expression
         :limit 5001}}
```

new style after this change:

```clojure
{:database 122
 :type :query
 :query {:source-table "card__1715"
         :breakout [[:field 488 {:base-type :type/Integer}]
                    [:field "full name" {:base-type :type/Text}]]
         :limit 5001}}
```

dpsutton committed:
Normalize field references

The schema was expecting valid mbql field refs (aka vectors and
keywords) but was getting a list and string (`[:field 23 nil]`
vs ("field" 23 nil)`). So normalize the field ref so we can handle the
stuff over the wire

this nice little bit of normalization lived in models.interface and
comped two functions in mbql.normalize. A few places reused it so moved
it to the correct spot.

dpsutton committed:
Better error message in `fix-expression-refs`

handles [:field ...] and [:expression ...] clauses. Seems unlikely that
aggregations will flow through here as that's not a great way to label a
pk. But i'll add support for that in a follow up

dpsutton committed:
Move `normalize-field-ref` below definition of `normalize-tokens`

`normalize-tokens` is `declare`d earlier, but we aren't using the var as
a var, but in a def we derefence it. But that's nil as it hadn't been
defined yet. Caused lots of failures downstream, including in cljs land

```clojure
user=> (declare x)
user=> (def y x) ;; <- the use in a def gets its current value
user=> (def x 3) ;; <- and it's not reactive and backfilling
user=> y
```

dpsutton committed:
Don't capture `declare`d vars in a def

need late binding. when the function is called, those vars will have a
binding. But if we use a `def`, it grabs their current value which is an
unbound var.
